Don't require PremiumLevel; allow general premium levels

jamesgary opened the pull request:
See https://github.com/NoRedInk/noredink-ui/issues/85

This makes the `PremiumConfig` more general and not rely on a specific `PremiumLevel`. It also now requires `isFree : level -> Bool` and `allowedFor : level -> level -> Bool`.

The upside is that it's no longer coupled to any specific `PremiumLevel` and can we can change/add the levels in the monolith. The downside is that all premium checkboxes need the new `isFree` and `allowedFor` configs, but those should probably be defined functions in `Data.PremiumLevel` so there's not too much duplication of code.

Feedback welcome! /cc @tesk9 @SamGerber 

Side note: Not sure if this will need to get bumped from `Nri.Ui.PremiumCheckbox.V1` to `Nri.Ui.PremiumCheckbox.V2` since it's not getting used in the monolith. If it does need to get bumped, I'll keep `V1` as-is and move changes to `V2`.

avh4 commented on src/Nri/Ui/PremiumCheckbox/V1.elm:
```diff
@@ -22,13 +21,15 @@ import Nri.Ui.Data.PremiumLevel as PremiumLevel exposing (PremiumLevel(..))
+    , teacherPremiumLevel : level
+    , contentPremiumLevel : level
+    , isFree : level -> Bool
+    , allowedFor : level -> level -> Bool
```
💭 This is a case where I think it's very risky to not label the parameters, since flipping them will case runtime bugs.  Maybe something like `allowedFor: { userLevel : level, checkboxContentLevel : level } -> Bool` ?

avh4 commented on src/Nri/Ui/PremiumCheckbox/V1.elm:
```diff
@@ -22,13 +21,15 @@ import Nri.Ui.Data.PremiumLevel as PremiumLevel exposing (PremiumLevel(..))
-    , teacherPremiumLevel : PremiumLevel
-    , contentPremiumLevel : PremiumLevel
+    , teacherPremiumLevel : level
+    , contentPremiumLevel : level
```
💭 hmm,  if we're passing the levels just so the checkbox can pass them to functions that we are also passing, then why not just pass the bools `isFree : Bool, isAllowed : Bool` instead of passing the levels and the functions?

jamesgary commented:
Simplified the config to just use two booleans: `isFree` and `isLocked`. I also noticed that these should never both be true; why would something free be locked? So I introduced a new type in a [different commit](https://github.com/NoRedInk/noredink-ui/commit/8d0ef9b26f9004c45a28f6e572c294023160a30a) outside of this PR: 

```
type Status
    = Free
    | PaidAndLocked
    | PaidAndUnlocked
```

But this didn't seem to improve things much. It made the module and client code more verbose, and offloaded the work to determine the `Status` to the client, which doesn't make it more convenient. I think I'm fine sticking with the two booleans approach.

jamesgary commented:
Ready for another review! 

- Added the ability to show a `P+` pennant
- Simplified the `PremiumConfig` a bit (no more `showFlagWhenLocked: Bool` since I didn't see that logic in prod anywhere, so I just have `pennant: Maybe Pennant`.
- Simplified example. The original controls didn't align with how the API is getting used, so I just made a handful of examples instead of dynamic controls. I can add them back in if we still think they're useful though.
- Added `.iconPremiumWithWritingFlag_svg` to `Assets`, which is the same name that's in prod

Side note (from above): Not sure if this will need to get bumped from Nri.Ui.PremiumCheckbox.V1 to Nri.Ui.PremiumCheckbox.V2 since it's not getting used in the monolith. If it does need to get bumped, I'll keep V1 as-is and move changes to V2.

SamGerber commented on src/Nri/Ui/PremiumCheckbox/V1.elm:
```diff
@@ -27,29 +26,29 @@ type alias PremiumConfig msg =
 
+{-| Premium is the yellow "P" pennant
+PremiumWithWriting is the yellow "P+" pennant
+-}
```
😻

SamGerber commented on src/Nri/Ui/PremiumCheckbox/V1.elm:
```diff
@@ -27,29 +26,29 @@ type alias PremiumConfig msg =
+
 {-| A checkbox that should be used for premium content
 
 This checkbox is locked when the premium level of the content is greater than the premium level of the teacher
```
I can't decide if this comment is helpful here any longer.
